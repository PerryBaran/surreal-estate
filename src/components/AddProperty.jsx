import React, { useReducer } from "react";
import style from "../styles/addProperty.module.css";

const initialState = {
  title: "",
  type: "Flat",
  bedrooms: 1,
  bathrooms: 1,
  price: 0,
  city: "Manchester",
  email: "",
};

const reducer = (state, { type, key, value }) => {
  switch (type) {
    case "update":
      return {
        ...state,
        [key]: value,
      };
    default:
      return {
        state,
      };
  }
};

const AddProperty = () => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const handleAddProperty = (e) => {
    e.preventDefault();
    console.log(state);
  };

  const handleFieldChange = (e) => {
    e.preventDefault();
    const { value, name } = e.target;
    dispatch({
      type: "update",
      key: name,
      value,
    });
  };

  return (
    <div className={style["add-property"]}>
      <form onSubmit={handleAddProperty} className={style.form}>
        <label htmlFor="title" className={style.label}>
          Title:
          <input
            id="title"
            name="title"
            value={state.title}
            onChange={handleFieldChange}
            placeholder="2 bed flat"
            className={style.input}
          />
        </label>
        <label htmlFor="type">
          Type:
          <select
            id="type"
            name="type"
            value={state.type}
            onChange={handleFieldChange}
            className={style.input}
          >
            <option value="Flat">Flat</option>
            <option value="Detached">Detached</option>
            <option value="Semi-Detached">Semi-Detached</option>
            <option value="Terraced">Terraced</option>
            <option value="End Of Terrace">End Of Terrace</option>
            <option value="Cottage">Cottage</option>
            <option value="Bungalow">Bungalow</option>
          </select>
        </label>
        <div className={style.rooms}>
          <label htmlFor="bedrooms">
            Bedrooms:
            <input
              type="number"
              id="bedrooms"
              name="bedrooms"
              value={state.bedrooms}
              min={1}
              max={9}
              onChange={handleFieldChange}
              className={style.number}
            />
          </label>
          <label htmlFor="bathrooms">
            Bathrooms:
            <input
              type="number"
              id="bathrooms"
              name="bathrooms"
              value={state.bathrooms}
              min={1}
              max={9}
              onChange={handleFieldChange}
              className={style.number}
            />
          </label>
        </div>
        <label htmlFor="price">
          Price:
          <input
            type="number"
            id="price"
            name="price"
            value={state.price}
            min={0}
            step="any"
            onChange={handleFieldChange}
            className={style.input}
          />
        </label>
        <label htmlFor="city">
          City:
          <select
            id="city"
            name="city"
            value={state.city}
            onChange={handleFieldChange}
            className={style.input}
          >
            <option value="Manchester">Manchester</option>
            <option value="Leeds">Leeds</option>
            <option value="Sheffield">Sheffield</option>
            <option value="Liverpool">Liverpool</option>
          </select>
        </label>
        <label htmlFor="email">
          Email:
          <input
            type="email"
            id="email"
            name="email"
            value={state.email}
            onChange={handleFieldChange}
            placeholder="perry.baran@email.com"
            className={style.input}
          />
        </label>
        <button type="submit">Add</button>
      </form>
    </div>
  );
};

export default AddProperty;
