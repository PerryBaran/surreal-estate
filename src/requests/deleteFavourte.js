/* eslint-disable no-console */
import axios from "axios";
import endpoint from "../data/endpoint";

const deleteFavourite = async (favouriteId, setAlert) => {
  const address = `${endpoint}/Favourite/${favouriteId}`;

  try {
    await axios.delete(address);
    setAlert("");
  } catch (err) {
    console.error(err);
    setAlert(
      "Server error, failed to remove favourite. Please try again later."
    );
  }
};

export default deleteFavourite;
