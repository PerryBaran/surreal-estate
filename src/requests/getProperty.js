/* eslint-disable no-console */
import axios from "axios";
import endpoint from "../data/endpoint";

const getProperty = async (setProperties, setAlert, search) => {
  let address = `${endpoint}/PropertyListing`;
  if (search) {
    address += search;
  }

  try {
    const { data } = await axios.get(address);
    setProperties(data);
    setAlert("");
  } catch (err) {
    console.error(err);
    setProperties([]);
    setAlert("Server Error, please try again later.");
  }
};

export default getProperty;
