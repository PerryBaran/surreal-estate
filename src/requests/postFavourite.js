/* eslint-disable no-console */
import axios from "axios";
import endpoint from "../data/endpoint";

const postFavourite = async (propertyInfo, setTriggerGet) => {
  const address = `${endpoint}/Favourite`;

  try {
    await axios.post(address, propertyInfo);
    setTriggerGet((prev) => prev + 1);
  } catch (err) {
    console.error(err);
  }
};

export default postFavourite;
