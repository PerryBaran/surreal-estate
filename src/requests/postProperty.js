import axios from "axios";
import endpoint from "../data/endpoint";

const postProperty = async (property) => {
  const address = `${endpoint}/PropertyListing`;

  try {
    await axios.post(address, property);
  } catch (err) {
    throw new Error(err);
  }
};

export default postProperty;
