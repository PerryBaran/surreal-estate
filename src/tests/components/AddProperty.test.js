import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import AddProperty from "../../components/AddProperty";
import * as postProperty from "../../requests/postProperty";

describe("AddProperty", () => {
  const validProps = {
    cities: ["Leeds", "Manchester"],
    types: ["Flat", "Detached"],
  };

  test("snapshot", () => {
    const { asFragment } = render(
      <AddProperty cities={validProps.cities} types={validProps.types} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("form tests", () => {
    beforeEach(() => {
      render(
        <AddProperty cities={validProps.cities} types={validProps.types} />
      );
    });

    test("content", () => {
      const title = screen.getByLabelText(/title/i);
      const typeSelect = screen.getByLabelText(/type/i);
      const bedrooms = screen.getByLabelText(/bedrooms/i);
      const bathrooms = screen.getByLabelText(/bathrooms/i);
      const price = screen.getByLabelText(/price/i);
      const citySelect = screen.getByLabelText(/city/i);
      const email = screen.getByLabelText(/email/i);

      expect(screen.getByRole("form")).toBeInTheDocument();
      expect(title).toHaveAttribute("placeholder", "2 bed flat");
      expect(typeSelect).toHaveDisplayValue(validProps.types[0]);
      validProps.types.forEach((type) => {
        const option = screen.getByText(type);

        expect(typeSelect).toContain(option);
        expect(option).toHaveAttribute("value", type);
      });
      expect(bedrooms).toHaveAttribute("type", "number");
      expect(bedrooms).toHaveDisplayValue("0");
      expect(bedrooms).toHaveAttribute("min", "0");
      expect(bathrooms).toHaveAttribute("type", "number");
      expect(bathrooms).toHaveDisplayValue("0");
      expect(bathrooms).toHaveAttribute("min", "0");
      expect(price).toHaveAttribute("type", "number");
      expect(price).toHaveDisplayValue("0");
      expect(price).toHaveAttribute("min", "0");
      expect(citySelect).toHaveDisplayValue(validProps.cities[0]);
      validProps.cities.forEach((city) => {
        const option = screen.getByText(city);

        expect(citySelect).toContain(option);
        expect(option).toHaveAttribute("value", city);
      });
      expect(email).toHaveAttribute("type", "email");
      expect(email).toHaveAttribute("placeholder", "fake.email@example.com");
    });

    describe("alerts", () => {
      test("missing title", () => {
        const submit = screen.getByText(/add/i);
        let alert = screen.queryByText(/please provide a valid title/i);

        expect(alert).not.toBeInTheDocument();
        fireEvent.click(submit);

        alert = screen.getByText(/please provide a valid title/i);

        expect(alert).toBeInTheDocument();
      });

      test("price must be more than 0", () => {
        const title = screen.getByLabelText(/title/i);
        const submit = screen.getByText(/add/i);

        let alert = screen.queryByText(/price must be more than 0/i);

        expect(alert).not.toBeInTheDocument();
        fireEvent.change(title, { target: { value: "title" } });
        fireEvent.click(submit);

        alert = screen.getByText(/price must be more than 0/i);

        expect(alert).toBeInTheDocument();
      });

      test("non valid email", () => {
        const title = screen.getByLabelText(/title/i);
        const price = screen.getByLabelText(/price/i);
        const submit = screen.getByText(/add/i);

        let alert = screen.queryByText(/please provide a valid email/i);

        expect(alert).not.toBeInTheDocument();
        fireEvent.change(title, { target: { value: "title" } });
        fireEvent.change(price, { target: { value: 20 } });
        fireEvent.click(submit);

        alert = screen.getByText(/please provide a valid email/i);

        expect(alert).toBeInTheDocument();
      });
    });
  });

  describe("mocked postProperty", () => {
    let mockPostProperty;

    beforeEach(() => {
      mockPostProperty = jest.spyOn(postProperty, "default");
    });

    test("on sucess", () => {
      mockPostProperty.mockResolvedValue();
      render(
        <AddProperty cities={validProps.cities} types={validProps.types} />
      );

      waitFor(() => {
        expect(screen.getByText(/property added/i)).toBeInTheDocument();
      });
    });

    test("on error", () => {
      const message = "error";
      mockPostProperty.mockRejectedValue({ message });

      render(
        <AddProperty cities={validProps.cities} types={validProps.types} />
      );

      waitFor(() => {
        expect(screen.getByText(message)).toBeInTheDocument();
      });
    });
  });
});
