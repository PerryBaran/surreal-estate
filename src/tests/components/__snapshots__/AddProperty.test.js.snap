// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty snapshot 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <form
      aria-label="form"
      class="add-property-form"
    >
      <label
        class="add-property-label"
        for="title"
      >
        <span>
          Title
        </span>
        <input
          class="add-property-input"
          id="title"
          name="title"
          placeholder="2 bed flat"
          required=""
          value=""
        />
      </label>
      <label
        class="add-property-label"
        for="type"
      >
        <span>
          Type
        </span>
        <select
          class="add-property-select"
          id="type"
          name="type"
        >
          <option
            value="Flat"
          >
            Flat
          </option>
          <option
            value="Detached"
          >
            Detached
          </option>
        </select>
      </label>
      <label
        class="add-property-label"
        for="bedrooms"
      >
        <span>
          Bedrooms
        </span>
        <input
          class="add-property-input"
          id="bedrooms"
          min="0"
          name="bedrooms"
          type="number"
          value="0"
        />
      </label>
      <label
        class="add-property-label"
        for="bathrooms"
      >
        <span>
          Bathrooms
        </span>
        <input
          class="add-property-input"
          id="bathrooms"
          min="0"
          name="bathrooms"
          type="number"
          value="0"
        />
      </label>
      <label
        class="add-property-label"
        for="price"
      >
        <span>
          Price
        </span>
        <div
          class="add-property-price"
        >
          <span>
            Â£
          </span>
          <input
            id="price"
            min="0"
            name="price"
            required=""
            step="any"
            type="number"
            value="0"
          />
        </div>
      </label>
      <label
        class="add-property-label"
        for="city"
      >
        <span>
          City
        </span>
        <select
          class="add-property-select"
          id="city"
          name="city"
        >
          <option
            value="Leeds"
          >
            Leeds
          </option>
          <option
            value="Manchester"
          >
            Manchester
          </option>
        </select>
      </label>
      <label
        class="add-property-label"
        for="email"
      >
        <span>
          Email
        </span>
        <input
          class="add-property-input"
          id="email"
          name="email"
          placeholder="fake.email@example.com"
          required=""
          type="email"
          value=""
        />
      </label>
      <button
        class="add-property-button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
